3.0/m_relaymsg.cpp

```diff
@@ -0,0 +1,204 @@
+        : Command(parent, "RELAYMSG", 3, 3)
+        , cap(Cap)
+        , captag(Captag)
+    {
+        flags_needed = 'o';
+        syntax = "<channel> <nick> <text>";
+        allow_empty_last_param = false;
+    }
+
+    std::string GetFakeHostmask(const std::string& nick) {
+        return InspIRCd::Format("%s!%s@%s", nick.c_str(), fake_ident.c_str(), fake_host.c_str());
+    }
+
+    CmdResult Handle(User* user, const CommandBase::Params& parameters)
+    {
+        const std::string& channame = parameters[0];
+        const std::string& nick = parameters[1];
+        const std::string& text = parameters[2];
+
+        // Check that the source has the relaymsg capability.
+        if (IS_LOCAL(user) && !cap.get(user)) {
+            return CMD_FAILURE;
+        }
+
+        Channel* channel = ServerInstance->FindChan(channame);
+        // Make sure the channel exists and the sender is in the channel
+        if (!channel)
+        {
+            user->WriteNumeric(Numerics::NoSuchChannel(channame));
+            return CMD_FAILURE;
+        }
+        if (!channel->HasUser(user))
+        {
+            user->WriteNumeric(Numerics::CannotSendTo(channel, "You must be in the channel to use this command."));
+            return CMD_FAILURE;
+        }
+
+        // Check that target nick is not already in use
+        if (ServerInstance->FindNick(nick))
+        {
+            user->WriteNumeric(ERR_BADRELAYNICK, nick, "RELAYMSG spoofed nick is already in use");
+            return CMD_FAILURE;
+        }
+
+        // Make sure the nick doesn't include any core IRC characters (e.g. *, !)
+        // This should still be more flexible than regular nick checking - in particular
+        // we want to allow "/" and "~" for relayers
+        if (nick.find_first_of("!+%@&#$:'\"?*,.") != std::string::npos)
+        {
+            user->WriteNumeric(ERR_BADRELAYNICK, nick, "Invalid characters in spoofed nick");
+            return CMD_FAILURE;
+        }
+
+        // If the sender was a lcoal user, check that the target includes a nick separator
+        if (IS_LOCAL(user))
+        {
+            if (nick.find_first_of(cap.nick_separators) == std::string::npos)
+            {
+                user->WriteNumeric(ERR_BADRELAYNICK, nick, InspIRCd::Format("Spoofed nickname must include one of the following separators: %s", cap.nick_separators.c_str()));
+                return CMD_FAILURE;
+            }
+        }
+
+        // Send the message to everyone in the channel
+        std::string fakeSource = GetFakeHostmask(nick);
+        ClientProtocol::Messages::Privmsg privmsg(fakeSource, channel, text);
+        // Tag the message as @relaymsg=<nick> so the sender can recognize it
+        privmsg.AddTag("relaymsg", &captag, user->nick);
+        channel->Write(ServerInstance->GetRFCEvents().privmsg, privmsg);
+
+        if (IS_LOCAL(user))
+        {
+            // Pass the message on to other servers
+            CommandBase::Params params;
```

Comment:
This won't pass along the message tags which were sent on the message. Is that intended?

If not you can probably do something like this:

```cpp
params.insert(parameters.GetTags().begin(), parameters.GetTags().end());

Reply:
Hmm, I didn't really consider this case - I'm not sure if there's any use-case at the moment for a bot to send message tags in the /relaymsg call itself.

With that snippet I get this error:

```
/home/james/inspircd3/src/modules/m_relaymsg.cpp: In member function ‘virtual CmdResult CommandRelayMsg::Handle(User*, const CommandBase::Params&)’:
/home/james/inspircd3/src/modules/m_relaymsg.cpp:157:83: error: no matching function for call to ‘CommandBase::Params::insert(insp::detail::flat_map_base<std::pair<std::__cxx11::basic_string<char>, ClientProtocol::MessageTagData>, std::greater<std::__cxx11::basic_string<char> >, std::__cxx11::basic_string<char>, insp::detail::map_pair_compare<std::pair<std::__cxx11::basic_string<char>, ClientProtocol::MessageTagData>, std::greater<std::__cxx11::basic_string<char> > > >::const_iterator, insp::detail::flat_map_base<std::pair<std::__cxx11::basic_string<char>, ClientProtocol::MessageTagData>, std::greater<std::__cxx11::basic_string<char> >, std::__cxx11::basic_string<char>, insp::detail::map_pair_compare<std::pair<std::__cxx11::basic_string<char>, ClientProtocol::MessageTagData>, std::greater<std::__cxx11::basic_string<char> > > >::const_iterator)’
  157 |             params.insert(parameters.GetTags().begin(), parameters.GetTags().end());
```

Reply:
(fixed below)